lib/crawler/output_sink/elasticsearch.rb

```diff
@@ -31,17 +32,21 @@ def initialize(config)
         # initialize client now to fail fast if config is bad
         client
 
+        @queue_locked = false
```

Comment:
is it thread-safe? Is crawler mutithreaded, if so it probably deserves a proper synchronisation primitive?

Reply:
So locks are per sink? How does it work with crawler crawling multiple sites (if it can)?

Reply:
I'm not familiar with the implementation though, you collect documents from 10 threads into this sink and at some point the sink blocks and send a batch request to Elasticsearch, right?

Reply:
> I'm not familiar with the implementation though, you collect documents from 10 threads into this sink and at some point the sink blocks and send a batch request to Elasticsearch, right?

Basically yeah, once the sink reaches the bulk queue's threshold (e.g. 100 docs) it should block to perform the indexing. During that time any attempts to add a doc to the queue should be blocked.

I had some alternative ideas for managing this that would take longer to implement. For example having the sink spawn threads that individually manage indexing, so the sink can always receive docs from the coordinator. I thought it might be overkill, though.

Reply:
Yeah I think current logic works well - you can indeed use some sort of lock/mutex abstraction here to make the @queue thread-safe. It will slow things down, but not slow enough IMO

Reply:
Fair enough, I'll try to implement something with mutex. Thanks both of you for flagging this!